### ISSUE TITLE
Frame overflows too common

### ISSUE THREAD

**Issue by alerque** [simoncozens/sile#620]
I think something might be wrong with our default penalties. At the very least we need to throw more warnings when we overflow the frame, but out of the box there are a surprising number of cases where lines just run into the margin and that's undesirable.

One case is the test for `tests/bug-269.sil`. We're overflowing the frame at the top of the second page and there is no reason in the world the default settings with these words and languages shouldn't be able to do better than that. Does this have to do with the language switch or is it something to do with that line having come from pushback when the footnote insertions happened?

![image](https://user-images.githubusercontent.com/173595/58108507-a27f5000-7bf4-11e9-9245-acc6ff38db19.png)

----

Another one I ran into is `tests/bug-301.sil`. In this case insertions are not involved and neither is pushback. Admittedly the line is tight and might not actually squish enough, but why didn't we get hyphenation at least?

![image](https://user-images.githubusercontent.com/173595/58149494-e06b8b00-7c6b-11e9-8090-c4e78024c235.png)

**Comment by alerque** [comment]
I might have found at least one cause of this. While working on replacing stdlib with Penlight in #682 I started to fail some tests that looked good to me. In fact some instances of the problem show here were actually working _correctly_ in the version where I'd refactored the object model. Working backwards it looks like there were some insidious bugs in the stdlib code (both caused by stdlib itself and in the usage) causing side effects.

```lua
local ObjA = ObjB + ObjC
```

A reasonably correct value was achieved for ObjA, but with a side effect of the values of ObjB also being changed. Most of the time this didn't matter since only ObjA was used in the end, but sometimes (for example in `pagebuilder.findBestBreak()` it would occasionally decide not to use ObjA and go back to using ObjB — which now has bad properties.

Proper object models and methods (addition of objects returning a new object rather than returning a modified original) will at least mitigate this. I don't know that all instances of this bug were due to that though, so this should be reviewed after that work is complete to see if it is still an issue.

**Comment by alerque** [comment]
Fixing the side effect math bugs did fix a number of cases that looked like iterations of this bug, but overall the measurement system refactoring has not made the bug go away. If anything it has made worse. But it has also made it more apparent what's happening.

Basically the current `master` version of SILE is being lenient about math in one direction but not the other. Theoretically when divvying up *stretch* and *shrink* values across a line, no instance should be assigned a ratio of greater than 1. Doing so would allow output that defies the input specifications. If `5pt plus 3pt minus 2pt` is allowed to grow beyond 8 points or shrink below 2 points, we're not handling ratios properly.

That's exactly what's being allowed to happen currently, but only in the stretch direction. Theoretically we have a `linebreak.emergencyStretch` parameter that controls how much we're allowed to fudge, and it _is_ being considered when weighting the penalties. But still we allow ourselves to go outside of those parameters without any warning to the user.

* In the stretch direction we just let it stretch no matter what the ratio ends up being. Space skip settings of `5pt plus 3pt minus 2pt` can result in spaces 100 points wide if need pee (set `typesetter.parfillskip` to 0 if you don't believe me).
* In the shrink direction we just stop shrinking and slobber all over the canvas even if that means going outside of the frame.

I propose the proper solution is twofold:

1. Introduce a warning when this happens. We notify for under and over-full frames in the frame advance direction, we should do the same for the line advance direction. When I render a book I shouldn't have to look through ever page in the PDF too see if there are overflows. Due to this bug I've literally been doing this by opening the PDF, blocking off the the allowed content area from my vision with a paper or another window, and holding PageDown to flip through all pages, if anything flickers on the screen in the margin area something is out of bounds. If so, then I know I need to tweak space or font metrics and try again ­— but this process needs to be repeated every time as minor content changes can throw everything off. Instead the console should warn me with a message that there were overflows.

2. Add a setting that toggles what to do. The current mix-match behavior where out of range stretches happen silently so there are never under-fills but out of range shrinks are rejected and over-fills happen instead should go away. Instead the out of range allowance (line ratios greater than 1 or less than -1) should both either allowed or capped.

   A possible 2b part to this would be to either add an "emergencyShrink" parameter to match "emergencyStretch" or change the existing parameter from a measurement to a length with both stretch and shrink values.

**Comment by alerque** [comment]
See also an example of similar problems in Urdu: https://groups.google.com/d/topic/sile-users/XR-NvBYZmhY

**Comment by Omikhleia** [comment]
As far as I can say (esp. on `tests/bug-301.sil` on page one, second line), the mere issue at stakes here is that the line:
- cannot be stretched or shrinked more within the defined constraints (in that classical scenario, the only elements that can be shrinked or stretched are the spaces).
- does not have any suitable hyphenation point ("should" cannot be broken).

The "workaround" here is simple - allow less than optimal stretching/shrinking. E.g. with `linebreak.tolerance` set to 2000 (instead of the default 500):
![image](https://user-images.githubusercontent.com/18075640/161828280-92731d10-46bb-4d3a-be07-d889242cd833.png)

That's still reasonable for a line that cannot possibly be made better.

So as in your first point above, @alerque, the only thing I'd regret here, compared to TeX, is the latter indeed warns the user (based on another parameter, a badness) when the amount of overflow is that bad (as it was the case here) and also has the option to output a fairly visible black box at the end of the line, to help reviewers. -- it can be enabled with the "draft" class option or at convenience, setting the `\overfullrule` length explicitly).

But other than that, there's not really a bug here. And the algorithms already has a setting for this (tolerance), so I am not sure we'd need anything else.

As of `tests/bug-269.sil`, I would say it's possibly the same thing. With a relaxed tolerance of 2000:

![image](https://user-images.githubusercontent.com/18075640/161829853-2a7b436a-553f-4d42-9819-8ece5ff0cefa.png)

But I have not idea here if the "müjde" word causing the issue could rather have been hyphenated. The TeX/hunspell hyphenation patterns do not show any breakpoint here, so it's not SILE's fault. But these patterns are not always complete... So one could add some hyphenation breakpoints, if this is acceptable. AFAIC, I do it quite often for neologism :laughing: 

As for the default tolerance, 500 is "usually" quite ok in languages with short words and/or words with easy hyphenation patterns. For French, I have nearly always needed a tolerance between 1000 and 2000 on book-like pages. I haven't check how LaTeX's babel behaves nowadays, but I remember the French packages by Bernard Gaulle  had a `\moretolerance` convenience command to double the value each time it's called (doh!).

**To recap my long text**
1. Being warned and having a way to visibly show the horizontal overflows could indeed be handy in some cases. 
2. The usual solution to these is to relax (temporarily or not) the `linebreak.tolerance`.

**Comment by Omikhleia** [comment]
Additional comment:
Of course instead of relaxing the tolerance, one can also tune the emergencyStretch, e.g.
```
%\set[parameter=linebreak.tolerance,value=2000]
\set[parameter=linebreak.emergencyStretch,value=15%lw]
```
(BTW it's a measurement, and should be relative most of the time to the line width rather than some fixed value)

The emergencyStretch has the advantage that it has no effect on good paragraphs. It is only applied if the paragraph would otherwise have failed to be within the tolerance, so it often gives better results.

As for imagining other strategies than TeX here...
- Capping the ratio is pretty bad (if there's an overflow on a line, it's _not_ where it should be handled by capping, the decision has to be paragraph-wise).
- Trying to toy with new concepts (emergencyShrink), while not necessarily plain wrong, implies deeply understanding the consequences (the emergencyStretch was added to TeX3 after much thought. Other than that:
  - As noted, a possible way of warning/showing line overflows
  - There's a fairly long analysis of emergencyStretch here https://tug.org/TUGboat/tb38-1/tb118wermuth.pdf - I wonder if we could make any use of it, e.g. to "compute" the ideal stretch for a bad paragraph and kick than that value in an additional iteration of the linebreaking algorithm.

**Comment by Omikhleia** [comment]
Also likely relates to #1404

**Comment by alerque** [comment]
Interestingly I'm not sure *any* of the examples show here are actually just problems with the defaults allowing overflows. They are pushback bugs or other problematic bits, but #1755 has far more concrete examples of actual "common" overflows that could have been something else.
